mkosi/machine.py

```diff
@@ -0,0 +1,97 @@
+# SPDX-License-Identifier: LGPL-2.1+
+
+from __future__ import annotations
+
+import contextlib
+import signal
+import sys
+from typing import Any, Optional, Sequence
+
+import pexpect  # type: ignore
+
+import mkosi
+
+from .backend import die
+
+
+class Machine():
+    def __init__(self, args: Optional[Sequence[str]] = None) -> None:
+        # Remains None until image is built and booted, then receives pexpect process
+        self.serial: pexpect.spawnu
+        self.stack = contextlib.ExitStack()
+        self.args = args
+        self.exit_code: int
+        self.verb: str
+
+
+    def __enter__(self) -> Machine:
+        with contextlib.ExitStack() as stack:
+            parsed_args = mkosi.parse_args(self.args)
+            parsed_args = parsed_args["default"]
+            loaded_args = mkosi.load_args(parsed_args)
+
+            self.verb = loaded_args.verb
+
+            mkosi.prepend_to_environ_path(loaded_args.extra_search_paths)
+
+            if loaded_args.verb in mkosi.MKOSI_COMMANDS_SUDO:
+                mkosi.check_root()
+                mkosi.unlink_output(loaded_args)
+
+            if loaded_args.verb == "build":
+                mkosi.check_output(loaded_args)
+
+            if mkosi.needs_build(loaded_args):
+                mkosi.check_root()
+                mkosi.check_native(loaded_args)
+                mkosi.init_namespace(loaded_args)
+                manifest = mkosi.build_stuff(loaded_args)
+
+                if loaded_args.auto_bump:
+                    mkosi.bump_image_version(loaded_args)
+
+                mkosi.save_manifest(loaded_args, manifest)
+
+                mkosi.print_output_size(loaded_args)
+
+            cmdline = None
+            if loaded_args.verb in ("shell", "boot"):
+                cmdline = mkosi.run_shell_cmdline(loaded_args)
+            elif loaded_args.verb == "qemu":
+                # We must keep the temporary file opened at run_qemu_cmdline accessible, hence the context stack.
+                cmdline = stack.enter_context(mkosi.run_qemu_cmdline(loaded_args))
+            else:
+                die("No valid verb was entered.")
+
+            cmdline = [str(x) for x in cmdline]
+            cmdline = " ".join(cmdline)
+            self.serial = pexpect.spawnu(cmdline, logfile=sys.stdout, timeout=240)
+            self.stack = stack.pop_all()
+
+        return self
+
+
+    def expect_character(self) -> None:
+        self.serial.expect("#")
+
+
+    def run(self, cmdline: str) -> None:
+        self.expect_character()
+        self.serial.sendline(cmdline)
+
+
+    def poweroff(self) -> None:
+        if self.verb in ("boot", "qemu"):
+            self.run("systemctl exit")
+        if self.verb == "shell":
+            self.run("exit")
+
+        self.serial.expect(pexpect.EOF)
+
+
+    def __exit__(self, *args: Any, **kwargs: Any) -> None:
+        self.poweroff()
+        self.serial.kill(signal.SIGTERM)
+        self.stack.__exit__(*args, **kwargs)
```

Comment:
Calling this immediately means we don't wait for the image to exit completely before we delete any temporary files managed by the context stack. This should be called after `self.serial.wait()`